### PR TITLE
Introduce configured_state arg for accelerator_config

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

This PR starts to allow users to bring in their own `Accelerator`. To start, we are simply allowing users to define a `PartialState` or `AcceleratorState` outside the `TrainingArguments`, and then enable the user to use its state using a new `use_configured_state` arg.

For instance, a user can now do:

```python
from accelerate import PartialState
from transformers import TrainingArguments

state = PartialState()
args = TrainingArguments(accelerator_config={"use_configured_state":True})
```

And this will use the defined `state` already made.

These `states` are Singeltons, so defining it once and calling it will maintain the same state on subsequent calls. 

This *may* lead to issues with hyperparameter tuning, which requires the state to be reset each time, as noted in the doc description

Fixes related accelerate issue which occurs when defining an `Accelerator` before the `TrainingArguments`: https://github.com/huggingface/accelerate/issues/2564

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@amyeroberts @pacman100

### CODE REVIEW

**Review by pacman100 — APPROVED** [review]
Thank you for adding `use_configured_state` to control whether or not to reset the `AcceleratorState` when creating `TrainingArguments` object!

File: src/transformers/training_args.py
Comment (amyeroberts) [inline src/transformers/training_args.py:532]: It would be good to flag that it might lead to issues with hyperparameter tuning here

File: src/transformers/training_args.py
Comment (amyeroberts) [inline src/transformers/training_args.py:1552]: Why do we need to add the `self.framework == "pt" and is_torch_available()` to the condition here?

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! 

Could we run a check with a hyperparam search (just most common settings) to check whether this does cause an effect?

File: src/transformers/training_args.py
Comment (muellerzr) [reply src/transformers/training_args.py:1552]: This particular block isn't under the torch-specific code, it can be moved there safely though I think

Comment (muellerzr) [reply src/transformers/training_args.py]: I've guarded this a bit better, let me know what you think!

**Comment by muellerzr** [comment]
cc @amyeroberts

File: src/transformers/training_args.py
Comment (amyeroberts) [inline src/transformers/training_args.py:1553]: I don't think we need to check if torch is available here for two reasons: 

1) it'd be confusing for a user to set `framework="pt"` and have this error show. Checking the framework vs. env compatibility should be the responsibility of another bit of the code 
2) This won't trigger if someone sets `framework="tf"` and they have torch installed in their env 

```suggestion
            if self.framework != "pt" :
```

File: tests/trainer/test_trainer.py
Comment (amyeroberts) [inline tests/trainer/test_trainer.py:2966]: Do we have to take the state from the global environment like this? It seems leaky and prone to possible issues. For example, how does this interplay with situations like multiprocessing? 

Could we instead pass in the state directly to the training arguments?

File: src/transformers/training_args.py
Comment (amyeroberts) [inline src/transformers/training_args.py:1955]: ultra-nit - we don't really need this line, as required behaviour is explain in the previous section. I don't mind if you want to leave it in

```suggestion
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for iterating on this! 

I have a question about the state specification, but I suspect this more to do with my knowledge of the use of `PartialState` in accelerate. 

Have you experimented at all with hyperparam searches to know if this does affect things?

File: tests/trainer/test_trainer.py
Comment (muellerzr) [reply tests/trainer/test_trainer.py:2966]: It's not from a global environment, the state is a stateful object that can only ever be initialized once. It's perfectly safe in a threaded environment. 

Ideally yes, we pass in something like the `PartialState` or `Accelerator` directly in instead, but it is very safe

Comment (amyeroberts) [reply tests/trainer/test_trainer.py:2966]: OK - sounds good to me!

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this! Looks good to me.

I'd like a second review and approval from @pacman100 before merging, as he's more familiar with the state handling so can catch anything I might have missed

**Review by pacman100 — APPROVED** [review]
Thank you @muellerzr for adding this!

**Comment by muellerzr** [comment]
@amyeroberts requesting a rereview after having to gut much of it on further testing 😅

**Comment by amyeroberts** [comment]
@pacman100 Can you do a re-review?

File: src/transformers/trainer_pt_utils.py
Comment (pacman100) [inline src/transformers/trainer_pt_utils.py:1315]: this parameter needs to be added to the docstrings.

File: src/transformers/training_args.py
Comment (pacman100) [inline src/transformers/training_args.py:1622]: ah, makes sense as the `PartialState` is usually created in the first call to `self.device` itself.

File: src/transformers/training_args.py
Comment (pacman100) [inline src/transformers/training_args.py:2050]: What would happen if `PartialState` already created is for MultiGPU while the user specifies `--use_cpu` in TrainerArguments? Currently, `--use_cpu` would be ignored. The same thing if PatialState is created for MultiGPU while user passes `--deepspeed` as in TrainerArguments.

**Review by pacman100 — COMMENTED** [review]
Thank you, @muellerzr, for iterating on this. Left a couple comments.

File: src/transformers/training_args.py
Comment (muellerzr) [reply src/transformers/training_args.py:2070]: Done!

File: src/transformers/training_args.py
Comment (pacman100) [inline src/transformers/training_args.py:2071]: this still doesn't account for the case when user passed `--fsdp` but hasn't enabled it via `PartialState`. In general, my comment was about the mismatch between the training arguments vs the PartialState set by them.

**Review by pacman100 — COMMENTED** [review]
Hello Zach, I still think this has some gaps. Please refer to my comment.

File: src/transformers/training_args.py
Comment (muellerzr) [reply src/transformers/training_args.py:2071]: If we just have a `PartialState` you can still initialize FSDP later as it's done at the `AcceleratorState` level. I'll do a quick test to verify, but the `PartialState` doesn't care about FSDP

Comment (muellerzr) [reply src/transformers/training_args.py:2071]: (PartialState only initializes the distributed env, for things like FSDP or DeepSpeed plugin that happens later, though DS will still be called/activated if the env is set properly with the PartialState)

Comment (muellerzr) [reply src/transformers/training_args.py:2071]: See here: https://github.com/huggingface/accelerate/blob/main/src/accelerate/state.py#L894-L904

Comment (pacman100) [reply src/transformers/training_args.py:2071]: Understood, Thank you Zach for the clarifications.

**Review by pacman100 — APPROVED** [review]

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this feature!

**Pull request merged**
